src/model/idtypes.rs

```diff
@@ -0,0 +1,299 @@
+use crate::model::Type;
+use serde::{Deserialize, Serialize};
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use strum::Display;
+use thiserror::Error;
+
+// This is a sealed trait pattern implementation, it stops external code from
+// implementing the `IdType` trait. The `Sealed` trait must be in a private mod,
+// so external code can not see and implement it.
+// See also: https://rust-lang.github.io/api-guidelines/future-proofing.html
+mod private {
+    pub trait Sealed {}
+}
+
+pub trait IdType: private::Sealed {
+    const TYPE: Type;
+}
+pub trait PlayableIdType: IdType {}
+pub trait PlayContextIdType: IdType {}
+
+macro_rules! sealed_types {
+    ($($name:ident),+) => {
+        $(
+            #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+            pub enum $name {}
+            impl private::Sealed for $name {}
+            impl IdType for $name {
+                const TYPE: Type = Type::$name;
+            }
+        )+
+    }
+}
+
+sealed_types!(Artist, Album, Track, Playlist, User, Show, Episode);
+
+impl PlayContextIdType for Artist {}
+impl PlayContextIdType for Album {}
+impl PlayableIdType for Track {}
```

Comment:
I'm a little bit confused about these two types, what's it for?

Reply:
These seem to be for generics, so that you can accept multiple types of related IDs (playable types such as tracks/episodes, etc).

Reply:
Yes, you are correct. Some methods accept only track/episode ids (`PlayableIdType`), and some methods operate on play contexts (`PlayContextIdType`). Hence these traits used to limit a set of id types to accept.